sonic-net/sonic-buildimage
# Add Secure Boot Support

#### Why I did it
Add Secure Boot support to SONiC OS.
Secure Boot (SB) is a verification mechanism for ensuring that code launched by a computer's UEFI firmware is trusted. It is designed to protect a system against malicious code being loaded and executed early in the boot process before the operating system has been loaded.
#### How I did it
Added a signing process to sign the following components:
shim, grub, Linux kernel, and kernel modules when doing the build, and when feature is enabled in build time according to the HLD explanations (the feature is disabled by default).
#### How to verify it
There are self-verifications of each boot component when building the image, in addition, there is an existing end-to-end test in sonic-mgmt repo that checks that the boot succeeds when loading a secure system (details below).

How to build a sonic image with secure boot feature: (more description in HLD)
1. Required to use the following build flags from rules/config:
- SECURE_UPGRADE_MODE="dev"
- SECURE_UPGRADE_DEV_SIGNING_KEY="/path/to/private/key.pem"
- SECURE_UPGRADE_DEV_SIGNING_CERT="/path/to/cert/key.pem"
After setting those flags should build the sonic-buildimage.
2. Before installing the image, should prepared the setup (switch device) with the follow:
- check that the device support UEFI
- stored pub keys in UEFI DB
- enabled Secure Boot flag in UEFI
3. How to run a test that verify the Secure Boot flow:
  The existing test "test_upgrade_path" under "sonic-mgmt/tests/upgrade_path/test_upgrade_path", is enough to validate proper boot
  You need to specify the following arguments:
  Base_image_list your_secure_image
  Taget_image_list your_second_secure_image
  Upgrade_type cold
  And run the test, basically the test will install the base image given in the parameter and then upgrade to target image by doing cold reboot and validates all the services are up and working correctly

note: the path to the existing sonic-mgmt test can be found in the link in the HLD description.
#### Which release branch to backport (provide reason below if selected)



#### Description for the changelog


#### Link to config_db schema for YANG module changes

No REDIS DB used in this feature

HLD link: https://github.com/sonic-net/SONiC/pull/1028

#### A picture of a cute animal (not mandatory but encouraged)
